Migration/l10n br account product

renatonlima opened the pull request.

renatonlima committed:
set l10n_br_account_product as installable

renatonlima committed:
update module metadata

renatonlima committed:
migrate l10n_br_account.fiscal.category object to new API

renatonlima committed:
migrate l10n_br_account.document.serie object to new API

renatonlima committed:
migrate l10n_br_account.partner.fiscal.type object to new API

renatonlima committed:
migrate l10n_br_account_product.cfop to new API

migrate l10n_br_account_product.cfop to new API

renatonlima committed:
migrate l10n_br_account_product.import.declaration and l10n_br_account_product.import.declaration.line to new API

renatonlima committed:
migrate product.template to new API

renatonlima committed:
migrate product.product to new API

renatonlima committed:
migrate res_company.py classes to new API

renatonlima committed:
migrate res_partner.py classes to new API

renatonlima committed:
set account_product_fiscal_classification.py to migrate to new API

renatonlima committed:
migrate account.payment.term to new API

renatonlima committed:
migrate account.tax to new API

renatonlima committed:
migrate account.invoice to new API

renatonlima committed:
migrate account.invoice.line to new API

renatonlima committed:
migrate account.invoice.tax to new API

renatonlima committed:
remove product.product view

renatonlima committed:
migrate product.template view

renatonlima committed:
update res.company view

renatonlima committed:
added ncm_id in fiscal.position.tax view

renatonlima committed:
update CFOP view

renatonlima committed:
update fiscal classification view

renatonlima committed:
migrate account.invoice and account.invoice.line view

renatonlima committed:
migrate account.invoice workflow

renatonlima committed:
migrate account invoice export wizard

renatonlima committed:
migrate nfe export object

renatonlima committed:
set nfe validation to migrate

renatonlima committed:
somes changes in txt serizalization method, maybe we can use pysped lib to refactoring this method

renatonlima committed:
migrate demo data

renatonlima committed:
migrate and update product demo data

renatonlima committed:
migrate security rules

renatonlima committed:
[FIX] rename field

renatonlima committed:
[REF] refactoring _compute_amount method

renatonlima committed:
[FIX] icms_value field definition

renatonlima committed:
[FIX] _validate_taxes and _get_tax_codes methods

renatonlima committed:
[FIX] tax compute method

renatonlima committed:
[FIX] nfe document

renatonlima committed:
fix nfe document

fix nfe document

fix nfe document

renatonlima committed:
fix nfe document txt export

renatonlima committed:
added txt validation di field

renatonlima committed:
fix domains in account.invoice fields_view_get

renatonlima committed:
[FIX] wrong field name in _serializer method

renatonlima committed:
[FIX] cnpj_cpf validate in fiscal document related

renatonlima committed:
[FIX] ie validate in fiscal document related

renatonlima committed:
added new item in PRODUCT_ORIGIN

renatonlima committed:
Added FCI

renatonlima committed:
added etree import

renatonlima committed:
added import and change view in account.invoice

renatonlima committed:
Initial commit to add new feature tax estimate

renatonlima committed:
[FIX]

[FIX]

renatonlima committed:
[FIX] carrier ie and total taxes in nfe export

renatonlima committed:
added copy=False in some fields in account.invoice

renatonlima committed:
set default to active column

renatonlima committed:
added permission for objects

renatonlima committed:
added _compute_tax_estimate

added _compute_tax_estimate 2

added _compute_tax_estimate 3

renatonlima committed:
[FIX] amount estimate taxes

renatonlima committed:
[FIX] added icms origin in nfe xml export

renatonlima committed:
[FIX] delivery address in nfe and product origin

renatonlima committed:
added fci and vTotTrib in txt export

renatonlima committed:
fix nfe export

renatonlima committed:
[FIX] account invoice object and views

renatonlima committed:
[FIX] l10n_br_account_product object and view

renatonlima committed:
[FIX] res.company object

renatonlima committed:
[FIX] res.company vie

renatonlima committed:
[FIX] res.partner object and view

renatonlima committed:
added new column in cfop csv data file

renatonlima committed:
[FIX] nfe xml export

renatonlima committed:
[FIX] nfe txt export

renatonlima committed:
moved models and views to a folder as new pattern

renatonlima committed:
added new __init__.py in models

renatonlima committed:
added fields in account invoice form view

renatonlima committed:
added fields in account invoice supplier form view

renatonlima committed:
adicionado campo UFSaidaPais na tag exporta da nfe 310

rvalyi commented:
OMG! this one is such a hassle to fix the PEP 8...

danimaribeiro commented on l10n_br_account_product/models/account_invoice.py:
```diff
@@ -0,0 +1,979 @@
+        [('1', 'Normal'),
+         ('2', 'Complementar'),
+         ('3', 'Ajuste')], 'Finalidade da Emissão', readonly=True,
+        states={'draft': [('readonly', False)]}, default='1')
```
https://github.com/Trust-Code/l10n-brazil/commit/02ce1253678ead01f1c34df218d27c2a9f5e7f11

danimaribeiro commented:
File l10n_br_refund.py:
https://github.com/Trust-Code/l10n-brazil/commit/61bc5afe666ef38618c789b9bc751eb9f681ee76

danimaribeiro commented:
File l10n_br_refund.py:
https://github.com/odoo-brazil/l10n-brazil/issues/25
https://github.com/Trust-Code/l10n-brazil/commit/b81197118f5d102dd1a58b8e7ed3c8af4d2515f8

sadamo committed:
[IMP] added date_maturity in invoice page "Vencimentos"

renatonlima committed:
fixed rebase glitch in account_invoice.py

rvalyi commented:
So I just rebased on 8.0. Unfortunately that wasn't trivial: there were several merge errors that I fixed by forcing the file of this initial PR at the failed given revision. Still at the end of the process I had this remaining diff that I decided to fix as an additional commit https://github.com/akretion/l10n-brazil/commit/e5aecbd41ae972fde6b6b935ccd7b66bd3ec587d

If this is needed, the branch of the PR before my rebase is available as a copy in https://github.com/rvalyi/l10n-brazil/tree/migration/l10n_br_account_product

@renatonlima I'll have a quick PEP-8 round but that would be cool if you could commit the tests after that...

rvalyi committed:
fixed PEP 8 in license headers

rvalyi committed:
fixed unused imports

rvalyi committed:
PEP 8 - WIP

mileo commented:
http://iws.ibpt.org.br/Help

renatonlima committed:
[MERGE]

renatonlima committed:
create test files in l10n_br_account_product and move country bc code from l10n_br_data_base to l10n_br_base

renatonlima committed:
fix onchange method in account.invoice and change tests

rvalyi committed:
autopep8 without --aggressive option

rvalyi committed:
serializer/txt.py: autopep8 --aggressive --aggressive

rvalyi committed:
models: autopep8 without --aggressive option

rvalyi committed:
sped: autopep8 without --aggressive option

rvalyi committed:
pep8 wip

pep8 wip

rvalyi commented:
so that is rebased onto 8,0 and with half the pep8 fixed. still the other half to fix in the txt and pysped mappings...

mileo committed:
[REF] Flake8 pass txt.py

Conflicts:
	l10n_br_account_product/sped/nfe/validator/txt.py

renatonlima committed:
[fix] remove unused variable in account.tax object compute_all method

rvalyi committed:
pep8

rvalyi committed:
pep8 wizard cleared

renatonlima committed:
added more tests

renatonlima committed:
[fix] onchange product_id in invoice line

renatonlima committed:
added demo data to improve tests

added demo data to improve tests

mileo committed:
[REF] Flake8 pass document.py

mileo committed:
[REF] Change in state selection using selection_add

mileo committed:
[MIG] Make l10n_br_account_product installable

mileo committed:
[FIX] l10n_br_account_product_fiscal_classification_form

mileo committed:
[MIG] Backport OCA view of account.product.fiscal.classification.template

rvalyi committed:
models: pep8 cleared

fixed wrong linebreaks from previous pep8 cleanup

fixed wrong string linebreaks from previous pep8 cleanup

pep8 cleared

rvalyi committed:
merge from Akretion prod

rvalyi committed:
pep8

pylint: relative imports

renatonlima committed:
fix icmsst compute

mbcosta committed:
indIEDest fix; merge from Akretion prod 7ac054193140ec7bc23d590c74ad97a0c09e086a

rvalyi committed:
pylint WIP, versionned as 8.0.1.0.0 before the NCM refactor in any case

rvalyi commented:
Please note that upcoming refactor from @renatonlima will need this patch in fiscal rules to make the test pass in the OCA configuration: https://github.com/OCA/account-fiscal-rule/pull/33
@mileo @danimaribeiro  one more approval and we can merge it now...

rvalyi commented:
Note. So we got the fix OCA/account-fiscal-rule#33 merged that's good.

Now we would also need the port from @renatonlima of the fiscal classification template into the refactor from @legalsylvain  to be able to accomplish this PR:
https://github.com/OCA/account-fiscal-rule/pull/35/files

These are the last miles guys...

mbcosta committed:
[Correção] Movidos metodos map_tax do objeto res_partner para o account_fiscal_position

rvalyi commented:
In case you wonder we already have tests green locally with the NCM refactor (and yes tests really running). I'm just trying to sort out the NCM refactor PR tests before we submit the l10n-brazil one: https://github.com/OCA/account-fiscal-rule/pull/35

rvalyi commented:
@mileo @renatonlima @danimaribeiro guys we are fu#*ed for merging this today. We all approved @legalsylvain fiscal rules refactor some weeks ago, but it turns out that tests weren't running as we were expecting. That's @danimaribeiro who just spot the glitch.

As far as I can see now, the upstream fiscal classification tests are badly broken and this even before @renatonlima port of the templating system. In fact probably only the tests themselves are broken as our tests in l10n-brazil seem to run smoothly. But as for today, even if I were to spent 1 hour on that now to fix the tests in the upstream fiscal classification, that wouldn't be nice either to merge all of it in a hurry with no review. So we won't have that branch green today yet, I'm sorry about this. But in any case we all missed the broken tests weeks ago so now is a bad time for crying.

I told @renatonlima to push his refactoring regardless of the upstream. As far we tested, it works nice along with the pending @renatonlima PR on account-fiscal-rules, so work can still be contributed. Sorry again we are in this situation.

Please read: https://github.com/OCA/account-fiscal-rule/pull/35#issuecomment-155645548

rvalyi commented:
see proof of what I was telling here: https://github.com/OCA/account-fiscal-rule/pull/37

mileo commented:
=/

renatonlima committed:
fix import

renatonlima committed:
redefining tax.definition classes

renatonlima committed:
remove permissions

renatonlima committed:
pep8

renatonlima committed:
redefine inherit class

redefine inherit class

renatonlima committed:
pep8

renatonlima committed:
fix views

renatonlima committed:
redefining product template view

renatonlima committed:
redefining company view

renatonlima committed:
fix test

renatonlima committed:
fix demo data

renatonlima committed:
set auto_install false

renatonlima committed:
remove views

renatonlima committed:
fix demo data

renatonlima committed:
fix account.product.fiscal.classification views

renatonlima committed:
replace ncm_id for fiscal_classification_id, now this field in product.template is not a property field, we can use it

renatonlima committed:
fix invoice customer and supplier tests

renatonlima committed:
pep8

renatonlima committed:
change l10n_br_account_product dependency

renatonlima committed:
pep8

rvalyi commented:
OK so finally WE ARE GREEN on that one!!!!!!!!!!!!!

Remarks now:
-  please increment version number in module and README as there is a light migration to accomodate with the fiscal rule refactor.
- I'll squash a few pep 8 commits together. That being said, not much is squashable due to the paralelisation of the pep8 commits and the refactor commits.

rvalyi commented:
cc @mileo @danimaribeiro @mbcosta

renatonlima committed:
fix txt nfe export, added travis pysped dependency and added test file for l10n_br_account_product to text txt and xml export

mileo commented:
Vou começar a revisar

mileo commented:
- [ ] Sale Invoice -> Cliente 1 SP Contribuinte-> Product: Mouse  **Valor ICMS ST  = -2.52**
- [x] Exportação da NF-e em TXT, sem importação no aplicativo de SP;

cc @renatonlima @rvalyi @danimaribeiro

renatonlima commented:
@mileo 

O problema do ICMS ST é que antes ele só adicionava o imposto se tivesse preenchido o campo MVA, eu tinha removido este if no código, porque existe a cobrança da ST sem o MVA apenas com a diferencia da alíquotas (própria e o de destino), e tinha adicionado um if para conferir se o imposto tem um percentual, porque neste caso o produto tem o imposto ICMS ST 0% e ai elimina este erro.

Também vou conferir isso nos testes, pois em uma operação de venda os campos ICMS ST devem ser todos zerados.

renatonlima commented:
@mileo 

Sobre a rotina de exportação, eu estou fazendo um teste para gerar o arquivo de exportação, o resultado da rotina não e relevante, pois assim que este PR for feito eu vou fazer outro PR para fazer a exportação do TXT através do PySPED

rvalyi commented:
sadly tests added in last commit didn't pass. Also last commit message is not appropriate, we will change that with a --amend.

renatonlima commented:
I will add some tests for invoice refund and create one more invoice to check ICMS ST tax.

rvalyi commented:
ok we are green again kudos @renatonlima !
now test coverage is 64% which is not too bad. The output txt and xml are not tested yet but at least their generation works. We will test their output in other PR's to avoid holding this one.

I'll still squash some 5 commits as I said.

Also I'll add my +1 once we have 2 more tests:
- one more invoice witrh ST test
- a refund test.

@mileo @danimaribeiro you already pressed us to merge that PR even before passing the tests. So once we have these two more tests I'm assuming you are still okay for the merge and I'll proceed to speed up the merge of the downstream. So please in the case you would detect any major issue please shout now. Again, we will be reactive on the next PR's the point is making it easier for the work downstream.

mileo commented:
Hi, Can we wait until monday? Today i talk a lot about the edoc workflow in pycon br....
I really like to think about it and make some tests.
[]s

rvalyi commented:
@mileo as you prefer. Let's wait then. We will look at your edoc stuff BTW but it's just we need that merge before thinking new features.

mileo commented:
About the tests i really like to cover:

the ICMS CSTs:
- [ ] CST 00
- [ ] CST 10
- [ ] CST 20
- [ ] CST 30
- [ ] CST 40
- [ ] CST 41
- [ ] CST 50
- [ ] CST 51
- [ ] CST 60
- [ ] CST 90

Simples Nacional
- [ ] CST 100
- [ ] CST 101
- [ ] CST 102
- [ ] CST 103
- [ ] CST 201
- [ ] CST 202
- [ ] CST 203
- [ ] CST 300
- [ ] CST 400
- [ ] CST 500
- [ ] CST 900


IPI
- [ ] Tributado
- [ ] Não tributado
- [ ] Aquisição de ativo 
- [ ] Aquisição de ativo c/ frete

Refund:
- [ ] Refund ativo
- [ ] Refund normal

Chages in product fiscal category:
- [ ] Normal -> ST

Extras
- [ ] Freight, Other Costs, Insurance:

mileo commented:
@rvalyi I think that invoice workflow can be more simple or the core workflow. 

Then it's is better if we merge with this change.

danimaribeiro commented on l10n_br_account_product/models/account.py:
```diff
@@ -0,0 +1,278 @@
+                calculed_taxes += result_icmsst['taxes']
+
+        # Estimate Taxes
+        if fiscal_position and fiscal_position.asset_operation:
```
We should change this to the "ind_final" flag that came with nfe 3.10:
https://github.com/odoo-brazil/l10n-brazil/issues/120

danimaribeiro commented:
Any comments about the workflow last friday?

danimaribeiro commented:
Those tests, better to create issues to do it later.
If someone creates, can put my name in 4, 5 of them

rvalyi commented:
@mileo we would also like more tests. But frankly this is features we didn't have before and beyond the OCA requirements. So really we are all in favor, it's just we will do that in next PR's. As for the workflow I think it's the same, it will require a new migration step eventually for existing users and it's better to break things down. Merging this doesn't mean we cannot merge an improvement a few days later.

This PR is basically only:
- the v8 API migration
- meeting OCA standards (that are much strictier in v8 than in v7 anyway)
- adapting to @legalsylvain fiscal classification refactor since you, @danimaribeiro @renatonlima and me all approved it for being OCA 8.0
- improvements we Akretion had to deal with during these months to meet legal requirements in our projects, not more not less.

rvalyi commented:
@danimaribeiro we basically agree on the principle but had not time to review the PR yet. Friday was more of a business / community meeting and not a codesprint. Well I think both kind of events are required...

mileo commented:
About the tests yes it's better to do later with PR.

My concern about the workflow is normal "people" dealing with update if we change it.

rvalyi commented:
@mileo we can do the merge and not advertise it a lot yet and in just a few days we do the wkf change... The point with doing that merge allows us to parallelize the work on lots of downstream PR's like enables @danimaribeiro to work on standardizing the service electronic invoice etc...

Anyway not merging this right now, but as soon as you think it's better to move forward let's do it.

mileo commented:
:100: LGTM: We can merge now!
- Notify people in odoo list to test it and say also that  in just a few days we do the wkf change.
- Create the test issues; ( I can put 2 devs working on it this week );

renatonlima committed:
account invoice refund migration

renatonlima committed:
added demo data for sale refund

renatonlima committed:
added basic test for invoice customer refund

lfdivino commented:
@renatonlima 

Verifiquei um erro no teste account_supplier_invoice.yml

##   I check that the invoice state is "sefaz_export"

  !assert {model: account.invoice, id: account_invoice_supplier_sp_sp}:
    - state == 'open'

Saida >>> draft =! open

renatonlima committed:
fix product_fiscal_category_map

renatonlima committed:
added demo data for VendaST

mileo commented:
@renatonlima 

Why we don't use unittest2? Yml tests sucks! Can we debug it?

danimaribeiro commented:
@lfdivino The build is passing.

@mileo You can debug, put breakpoint in the methods yml calls.

rvalyi commented:
@mileo if you want to contribute additional tests in later PR's we will certainly accept unittest2 . Turns out that yml tests are handy for functional tests once you get used to it. But indeed they are harder to debug, but at least you need to do that only when they fail...

lfdivino commented:
@danimaribeiro 
Locally i have this errors: http://pastebin.com/cjQ1uG1B

mileo commented:
New tests: https://github.com/akretion/l10n-brazil/pull/5

mileo commented:
Residual, Freight and Other costs not working
Cherry pick need https://github.com/odoo-brazil/l10n-brazil/blob/8.0/l10n_br_account_product/account.py#L70

mileo commented:
@renatonlima Why this hash? fp_78df616ab31e95ee46c6a519a2ce9e12_internal_template

echo '?????' | md5sum

renatonlima committed:
fix bad file name

renatonlima committed:
fix and add demo data

renatonlima committed:
refactoring invoice onchanges and tax map

renatonlima committed:
fix demo data

renatonlima commented:
@mileo,

I comment your PR akretion#5, I agree this PR, but there is a change in other module (l10n_br_zip) at time I don't see problem, but please do next commits with related features.

renatonlima commented:
@lfdivino

I saw your logs and I think you are not using the corret branch or it is not update, please check our branch.

renatonlima commented:
@mileo,

The reason to use yml for those tests in l10n_br_account_product to test account.invoices is I took as a template the yml in account module.

mileo commented:
@renatonlima  About akretion#5, i use the 8.0 branch, with a pull of:
git pull github.com/akretion/l10n-brazil migration/l10n_br_account_product

The ZIP commit is already on OCA/8.0

renatonlima committed:
[MERGE]

rvalyi commented:
:+1: 
ok so I think we are finally clean enough on the migration for a merge. No feature is broken, we meet OCA standards and we have 67% of test coverage including this module which is by far the most complex of the localization. So okay for me and let's have further improvements in other PR's. As I said I'm squashing a few commits to keep a clean history.

mileo commented:
:+1:

mileo commented:
@renatonlima Please review the commit history of c4c66fc 
The autor is @lfdivino

rvalyi commented:
OK so I made a backup of the original branch here https://github.com/OCA/l10n-brazil/compare/8.0...akretion:migration/l10n_br_account_product-bk/
and from the 155 commits I squashed 18 commits from rvalyi and renatonlima. The logic applied is I squashed sequences (with no gap) of commits about the same concern that should have been just one (using amend possibly). The 18 commits I squashed are:

```
pick 3f2dd54 migrate l10n_br_account_product.cfop to new API
pick a0664b9 fix nfe document
pick a5e2609 fix nfe document
pick 6eb5b35 [FIX]
pick b773ad3 added _compute_tax_estimate 2
pick 1f44faa added _compute_tax_estimate 3
pick 7ce1baf pep8 wip
pick b3f3a8c fixed wrong linebreaks from previous pep8 cleanup
pick 9944511 fixed wrong string linebreaks from previous pep8 cleanup
pick 497ce7a pep8 cleared
pick 86a7895 added demo data to improve tests
pick 29d51a0 pylint: relative imports
pick 9badad6 redefine inherit class
pick 9b81bbe pep8
```

I think we can merge now then.
